EIPS/eip-721.md

```diff
@@ -0,0 +1,312 @@
+interface ERC721 {
+
+    // COMPLIANCE WITH ERC-165 (DRAFT) /////////////////////////////////////////
+
+    /// @dev ERC-165 (draft) interface signature for itself
+    // bytes4 internal constant INTERFACE_SIGNATURE_ERC165 = // 0x01ffc9a7
+    //     bytes4(keccak256('supportsInterface(bytes4)'));
+
+    /// @dev ERC-165 (draft) interface signature for ERC721
+    // bytes4 internal constant INTERFACE_SIGNATURE_ERC721 = // 0xe55729f5
+    //     bytes4(keccak256('ownerOf(uint256)')) ^
+    //     bytes4(keccak256('countOfDeeds()')) ^
+    //     bytes4(keccak256('countOfDeedsByOwner(address)')) ^
+    //     bytes4(keccak256('deedOfOwnerByIndex(uint256)')) ^
+    //     bytes4(keccak256('approve(address,uint256)')) ^
+    //     bytes4(keccak256('takeOwnership(uint256)'));
+
+    /// @notice Query a contract to see if it supports a certain interface
+    /// @dev Returns `true` the interface is supported and `false` otherwise,
+    ///  returns `true` for INTERFACE_SIGNATURE_ERC165 and
+    ///  INTERFACE_SIGNATURE_ERC721, see ERC-165 for other interface signatures.
+    function supportsInterface(bytes4 _interfaceID) external pure returns (bool);
+
+    // PUBLIC QUERY FUNCTIONS //////////////////////////////////////////////////
+
+    /// @notice Find the owner of a deed
+    /// @param _deedId The identifier for a deed we are inspecting
+    /// @dev Deeds assigned to zero address are considered destroyed, and
+    ///  queries about them do throw.
+    /// @return The non-zero address of the owner of deed `_deedId`, or `throw`
+    ///  if deed `_deedId` is not tracked by this contract
+    function ownerOf(uint256 _deedId) external view returns (address _owner);
+
+    /// @notice Count deeds tracked by this contract
+    /// @return A count of the deeds tracked by this contract, where each one of
+    ///  them has an assigned and queryable owner
+    function countOfDeeds() external view returns (uint256 _count);
+
+    /// @notice Count all deeds assigned to an owner
+    /// @dev Throws if `_owner` is the zero address, representing destroyed deeds.
+    /// @param _owner An address where we are interested in deeds owned by them
+    /// @return The number of deeds owned by `_owner`, possibly zero
+    function countOfDeedsByOwner(address _owner) external view returns (uint256 _count);
+
+    /// @notice Enumerate deeds assigned to an owner
+    /// @dev Throws if `_index` >= `countOfDeedsByOwner(_owner)` or if
+    ///  `_owner` is the zero address, representing destroyed deeds.
+    /// @param _owner An address where we are interested in deeds owned by them
+    /// @param _index A counter between zero and `countOfDeedsByOwner(_owner)`,
+    ///  inclusive
+    /// @return The identifier for the `_index`th deed assigned to `_owner`,
+    ///   (sort order not specified)
+    function deedOfOwnerByIndex(address _owner, uint256 _index) external view returns (uint256 _deedId);
```

Comment:
When an owner has 100 deeds, I need to call this function 100 times, but the owner might get or lose deeds during this period.  So, it's impossible to enumerate all deeds assigned to an owner, with this interface.

Reply:
Not if the call is made on-chain: even if your contract has to call `deedOfOwnerByIndex` 100 times, the result is guaranteed to be correct. The other contract can't be updated until your contract is done processing. Off-chain it can be an issue, but you can listen to Transfer events at the same time.

Reply:
Yup... But the risk onchain is to go out of gas. Implementing this feature is for sure gas expensive.

Reply:
The main use case we cater to is an off-blockchain wallet that wants to show you things that you own. (Possibly a browser extension having a logo of a fox that follows your cursor around.) I think the current standard is the best we can do, in terms of enumeration, for that standard. Even the note about O(1) being possible should be an inspiration, but I'm not sure we can standardize this. See also: CryptoKitties which [does NOT have](https://github.com/axiomzen/cryptokitties-bounty/blob/master/contracts/KittyOwnership.sol#L163-L179) an O(1) enumeration. Of course every implementation will have to make a memory / runtime trade off decision under the current standard.